src/Algebra/Ordered/Structures.agda

```diff
@@ -0,0 +1,205 @@
+
+  -- Preordered monoids (promonoids)
+
+  record IsPromonoid (_∙_ : Op₂ A) (ε : A) : Set (a ⊔ ℓ₁ ⊔ ℓ₂) where
+    field
+      isProsemigroup : IsProsemigroup _∙_
+      identity       : Identity _≈_ ε _∙_
+
+    open IsProsemigroup isProsemigroup public
+
+    isMonoid : IsMonoid _≈_ _∙_ ε
+    isMonoid = record
+      { isSemigroup = isSemigroup
+      ; identity    = identity
+      }
+
+    open IsMonoid isMonoid public using (identityˡ; identityʳ)
+
+  record IsCommutativePromonoid (_∙_ : Op₂ A) (ε : A) : Set (a ⊔ ℓ₁ ⊔ ℓ₂) where
+    field
+      isPromonoid : IsPromonoid _∙_ ε
+      comm        : Commutative _≈_ _∙_
+
+    open IsPromonoid isPromonoid public
+
+    isCommutativeMonoid : IsCommutativeMonoid _≈_ _∙_ ε
+    isCommutativeMonoid = record
+      { isMonoid        = isMonoid
+      ; comm            = comm
+      }
+
+  -- Preordered semirings (prosemirings)
+
+  record IsProsemiring (_+_ _*_ : Op₂ A) (0# 1# : A) : Set (a ⊔ ℓ₁ ⊔ ℓ₂) where
+    field
+      +-isCommutativePromonoid : IsCommutativePromonoid _+_ 0#
+      *-isPromonoid            : IsPromonoid _*_ 1#
+      distrib                  : _DistributesOver_ _≈_ _*_ _+_
+      zero                     : Zero _≈_ 0# _*_
+
+    distribˡ : _DistributesOverˡ_ _≈_ _*_ _+_
+    distribˡ = proj₁ distrib
+
+    distribʳ : _DistributesOverʳ_ _≈_ _*_ _+_
+    distribʳ = proj₂ distrib
+
+    zeroˡ : LeftZero _≈_ 0# _*_
+    zeroˡ = proj₁ zero
+
+    zeroʳ : RightZero _≈_ 0# _*_
+    zeroʳ = proj₂ zero
+
+    open IsCommutativePromonoid +-isCommutativePromonoid public
+      renaming
+      ( assoc       to +-assoc
+      ; ∙-cong      to +-cong
+      ; monotonic   to +-monotonic
+      ; identity    to +-identity
+      ; identityˡ   to +-identityˡ
+      ; identityʳ   to +-identityʳ
+      ; comm        to +-comm
+      ; isMagma     to +-isMagma
+      ; isSemigroup to +-isSemigroup
+      ; isMonoid    to +-isMonoid
+      )
+
+    open IsPromonoid *-isPromonoid public
+      using ()
+      renaming
+      ( assoc       to *-assoc
+      ; ∙-cong      to *-cong
+      ; monotonic   to *-monotonic
+      ; identity    to *-identity
+      ; identityˡ   to *-identityˡ
+      ; identityʳ   to *-identityʳ
+      ; isMagma     to *-isMagma
+      ; isSemigroup to *-isSemigroup
+      ; isMonoid    to *-isMonoid
+      )
+
+------------------------------------------------------------------------
+-- Partially ordered structures
+
+module _ {a ℓ₁ ℓ₂} {A : Set a}
+  (_≈_ : Rel A ℓ₁)  -- The underlying equality.
+  (_≤_ : Rel A ℓ₂)  -- The partial.
+  where
+
+  -- Partially ordered semigroups (posemigroups)
+
+  record IsPosemigroup (_∙_ : Op₂ A) : Set (a ⊔ ℓ₁ ⊔ ℓ₂) where
+    field
+      isPartialOrder : IsPartialOrder _≈_ _≤_
+      assoc          : Associative _≈_ _∙_
+      monotonic      : Monotonic₂ _≤_ _∙_
+
+    open IsPartialOrder isPartialOrder public
+
+    ∙-cong : Congruent₂ _≈_ _∙_
+    ∙-cong x≈y u≈v =
```

Comment:
You can use this pre-existing proof :smile: 

https://github.com/agda/agda-stdlib/blob/650e05f3a5ef06cb938c3b91a4fc8b9b8d7c5ef2/src/Relation/Binary/Consequences.agda#L89-L91

Reply:
Will do, thanks for pointing it out!

Reply:
Oops, that didn't work. The proof of `mono⇒cong` is for unary operations/functions, but `_∙_` is a binary operation. I see three ways to fix this:
 1. Leave the proof as is.

Reply:
If we move the definitions to `Relation.Binary` then I would vote for proposal 2.  (with the proofs implemented via proposal 3.!)

Reply:
I've now added a single new proof `mono₂⇒cong₂` to `Relation.Binary.Consequences`. It's independent of `Congruence` and `Monotonic` but it does *not* use `mono⇒cong` as that would require an additional assumption (transitivity) which the direct proof does not.